### PR TITLE
Remove IRowCursorConsolidator.

### LINKED ISSUE #1867: Remove IRowCursorConsolidator
Consider this method:

https://github.com/dotnet/machinelearning/blob/9067a1be58434999e37db7504d6b4e06eca8bbf4/src/Microsoft.ML.Core/Data/IDataView.cs#L114-L115

This returns an instance of the mysterious `IRowCursorConsolidator` interface. Why an interface? In retrospect I'm not quite sure.

Having an interface allows for different implementations, but we've never actually really exploited that capability. Nor, even if we were of such a mind to do so, would it be clear how we could. What would they even do differently? The semantics around `Batch` and whatnot are sufficiently clear and simple as to make only one implementation obvious, and even if we did have different implementations since the resulting cursors result most often from transformers (that is, components downstream from the set creation), they couldn't really do anything radically different anyway, since to do anything implementation specific on any cursor would be to break the composability at the core of what makes `IDataView` work at all.

So: get rid of this interface, and replace all usage of it with a simple utility method *somewhere* that can be called to do the reconciliation. It needn't even be a public utility method, but there may be reasons to do so.

### PR DESCRIPTION
**Pull request by TomFinley**
Fixes #1867.

### CODE REVIEW

File: src/Microsoft.ML.Core/Data/IDataView.cs [hunk 1]
```diff
@@ -122,20 +124,8 @@ public interface IDataView
...
-        /// <summary>
-        /// Create a consolidated cursor from the given parallel cursor set.
-        /// </summary>
-        RowCursor CreateCursor(IChannelProvider provider, RowCursor[] inputs);
-    }
-
-    /// <summary>
-    /// Delegate type to get a value. This can used for efficient access to data in an IRow
-    /// or IRowCursor.
+    /// Delegate type to get a value. This can used for efficient access to data in a <see cref="Row"/>
```
Comment (singlis) [inline src/Microsoft.ML.Core/Data/IDataView.cs]: minor - "can be used"?

File: src/Microsoft.ML.Core/Data/IDataView.cs [hunk 2]
```diff
@@ -146,43 +136,50 @@ public interface IRowCursorConsolidator
...
         public abstract long Position { get; }
 
         /// <summary>
-        /// This provides a means for reconciling multiple streams of counted things. Generally, in each stream,
-        /// batch numbers should be non-decreasing. Furthermore, any given batch number should only appear in one
-        /// of the streams. Order is determined by batch number. The reconciler ensures that each stream (that is
-        /// still active) has at least one item available, then takes the item with the smallest batch number.
+        /// This provides a means for reconciling multiple rows that have been produced generally from
+        /// <see cref="IDataView.GetRowCursorSet(Func{int, bool}, int, Random)"/>. When getting a set, there is a need
+        /// to, while allowing parallel processing to proceed, always have an aim thatthe original order should be
```
Comment (singlis) [inline src/Microsoft.ML.Core/Data/IDataView.cs]: minor - thatthe

File: src/Microsoft.ML.Core/Data/IDataView.cs [hunk 3]
```diff
@@ -146,43 +136,50 @@ public interface IRowCursorConsolidator
...
+        /// reconverable. So: for any cursor implementation, batch numbers should be non-decreasing. Furthermore, any
```
Comment (singlis) [inline src/Microsoft.ML.Core/Data/IDataView.cs]: recoverable?

Comment (singlis) [reply src/Microsoft.ML.Core/Data/IDataView.cs]: Are you saying that the original order should always be maintained?

Comment (TomFinley) [reply src/Microsoft.ML.Core/Data/IDataView.cs]: Given a cursor set, you should be (in principle) able to recover the order that would have been returned from a similarly configured call to the serial `GetRowCursor` by exploiting the `Batch` property, yes. I will perhaps try to make this more clear in the docs.

File: src/Microsoft.ML.Core/Data/IDataView.cs (hunk 2 above)
Comment (TomFinley) [reply src/Microsoft.ML.Core/Data/IDataView.cs]: Thanks @singlis !

File: src/Microsoft.ML.Core/Data/IDataView.cs (hunk 1 above)
Comment (TomFinley) [reply src/Microsoft.ML.Core/Data/IDataView.cs]: Yup, thank you!

File: src/Microsoft.ML.Core/Data/IDataView.cs (hunk 3 above)
Comment (TomFinley) [reply src/Microsoft.ML.Core/Data/IDataView.cs]: Whether I succeeded in clarifying though is another matter, if you could read over it and see if my few additional explanatory sentences help or hurt, I'd appreciate it, thanks @singlis !

File: src/Microsoft.ML.Data/Data/DataViewUtils.cs [hunk 4]
```diff
@@ -299,214 +294,198 @@ private Splitter(Schema schema)
...
                 }
+                int idIdx = activeToCol.Length + (int)ExtraIndex.Id;
+                outPipes[idIdx] = OutPipe.Create(NumberType.UG, GetPool(NumberType.UG, ourPools, idIdx));
 
-                private static RowCursor ConsolidateCore(IChannelProvider provider, RowCursor[] inputs, ref object[] ourPools, IChannel ch)
+                // Create the structures to synchronize between the workers and the consumer.
+                const int toConsumeBound = 4;
+                var toConsume = new BlockingCollection<Batch>(toConsumeBound);
+                var batchColumnPool = new MadeObjectPool<BatchColumn[]>(() => new BatchColumn[outPipes.Length]);
+                Thread[] workers = new Thread[inputs.Length];
```
Comment (singlis) [inline src/Microsoft.ML.Data/Data/DataViewUtils.cs]: So threads are used here, can the cursor set be large? Like in the hundreds for example? If so, would tasks be a better alternative?

**Comment by singlis** [comment]
Looks good to me - I didnt quite follow the "Scott document review".

File: src/Microsoft.ML.Data/Data/DataViewUtils.cs [hunk 5]
```diff
@@ -129,15 +129,13 @@ public static int GetThreadCount(IHost host, int num = 0, bool preferOne = false
...
-            var inputs = view.GetRowCursorSet(out consolidator, predicate, cthd, rand);
+            var inputs = view.GetRowCursorSet(predicate, cthd, rand);
             host.Check(Utils.Size(inputs) > 0);
-            host.Check(inputs.Length == 1 || consolidator != null);
 
             if (inputs.Length == 1)
                 curs = inputs[0];
             else
-                curs = consolidator.CreateCursor(host, inputs);
+                curs = DataViewUtils.ConsolidateGeneric(host, inputs, 64);
```
Comment (sfilipi) [inline src/Microsoft.ML.Data/Data/DataViewUtils.cs]: >64 [](start = 70, length = 2)

maybe have a comment for why 64.

**Pull request merged**